File: packages/picoipc/python/examples/roundtrip.py
#!/usr/bin/env python3
"""Single-process roundtrip: create ring, publish, consume."""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from smh_q import Ring


def main() -> int:
    name = f"smh_q_py_roundtrip_{os.getpid()}"
    Ring.unlink(name)

    cfg = dict(name=name, create=True, slot_count=8, slot_size=128, schema_id=42)
    producer = Ring(**cfg)
    consumer = Ring(name=name, create=False, schema_id=42)

    msg = b"hello-smh_q-py"
    assert producer.try_publish(msg), "publish failed"
    got = consumer.try_consume()
    assert got == msg, f"expected {msg!r}, got {got!r}"

    producer.close(unlink=True)
    print("roundtrip OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/picoipc/examples/cpp/roundtrip.cpp
#include "smh_q/ring.hpp"

#include <cstdio>
#include <cstring>
#include <string>
#include <unistd.h>

#define CHECK(cond)                                                          \
  do {                                                                       \
    if (!(cond)) {                                                           \
      std::fprintf(stderr, "CHECK failed: %s (%s:%d)\n", #cond, __FILE__,   \
                   __LINE__);                                                \
      std::abort();                                                          \
    }                                                                        \
  } while (0)

int main() {
  const std::string name = "smh_q_roundtrip_" + std::to_string(::getpid());
  smh_q::Ring::unlink(name);

  smh_q::Ring::Config cfg;
  cfg.name = name;
  cfg.slot_count = 8;
  cfg.slot_size = 128;
  cfg.schema_id = 42;

  {
    smh_q::Ring producer(cfg, true);
    smh_q::Ring consumer(cfg, false);

    const char* msg = "hello-smh_q";
    CHECK(producer.try_publish(std::span<const std::byte>(
        reinterpret_cast<const std::byte*>(msg), std::strlen(msg))));

    auto got = consumer.try_consume();
    CHECK(got.has_value());
    CHECK(got->size() == std::strlen(msg));
    CHECK(std::memcmp(got->data(), msg, got->size()) == 0);

    std::puts("roundtrip OK");
  }

  smh_q::Ring::unlink(name);
  return 0;
}
